server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -210,23 +214,27 @@ private Map<String, OriginalIndices> buildPerIndexOriginalIndices(
                 true,
                 indicesAndAliases
             );
-            BooleanSupplier hasDataStreamRef = () -> {
-                IndexAbstraction ret = clusterState.getMetadata().getIndicesLookup().get(index);
-                if (ret == null || ret.getParentDataStream() == null) {
-                    return false;
-                }
-                return indicesAndAliases.contains(ret.getParentDataStream().getName());
-            };
-            List<String> finalIndices = new ArrayList<>();
-            if (aliases == null || aliases.length == 0 || indicesAndAliases.contains(index) || hasDataStreamRef.getAsBoolean()) {
-                finalIndices.add(index);
+            String[] finalIndices = Strings.EMPTY_ARRAY;
```

Comment:
this bit here is the most questionable part to me. Seems like pre-allocating an Array list ( to aliase.length + 1) and then adding to it would be just as good and way more readable.

Reply:
`ArrayList` has a slower iterator later on, slower inserts because of the mode count update and takes more heap. I think I can do a follow-up to do away with building the list here anyway, having a non-resizable return is a good first step locking in a few things here :)